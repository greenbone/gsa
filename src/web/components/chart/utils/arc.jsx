/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {isDefined} from 'gmp/utils/identity';

import path from './path';

const EPSILON = 1e-12; // 1 * 10^(-12)

const PI2 = Math.PI * 2;

class Arc {
  constructor() {
    this._innerRadiusX = 0;
  }

  innerRadiusX(radius) {
    this._innerRadiusX = radius;
    return this;
  }

  outerRadiusX(radius) {
    this._outerRadiusX = radius;
    return this;
  }

  innerRadiusY(radius) {
    this._innerRadiusY = radius;
    return this;
  }

  outerRadiusY(radius) {
    this._outerRadiusY = radius;
    return this;
  }

  centroid({startAngle = 0, endAngle = PI2} = {}) {
    this._checkRadius();

    const outerRadiusX = this._outerRadiusX;
    const outerRadiusY = isDefined(this._outerRadiusY)
      ? this._outerRadiusY
      : outerRadiusX;

    const innerRadiusX = this._innerRadiusX;
    const innerRadiusY = isDefined(this._innerRadiusY)
      ? this._innerRadiusY
      : innerRadiusX;

    const rx = (innerRadiusX + outerRadiusX) / 2;
    const ry =
      isDefined(innerRadiusY) && isDefined(outerRadiusY)
        ? (innerRadiusY + outerRadiusY) / 2
        : rx;

    const a = (startAngle + endAngle) / 2;
    return {
      x: Math.cos(a) * rx,
      y: Math.sin(a) * ry,
    };
  }

  path({startAngle = 0, endAngle = PI2} = {}) {
    const paths = path();
    const diff = endAngle - startAngle;

    if (diff < EPSILON) {
      return paths.move(0, 0);
    }

    if (diff > PI2 - EPSILON) {
      endAngle = PI2;
    }

    this._checkRadius();

    const outerRadiusX = this._outerRadiusX;
    const outerRadiusY = isDefined(this._outerRadiusY)
      ? this._outerRadiusY
      : outerRadiusX;

    const innerRadiusX = this._innerRadiusX;
    const innerRadiusY = isDefined(this._innerRadiusY)
      ? this._innerRadiusY
      : innerRadiusX;

    const sx = outerRadiusX * Math.cos(startAngle);
    const sy = outerRadiusY * Math.sin(startAngle);
    const ex = outerRadiusX * Math.cos(endAngle);
    const ey = outerRadiusY * Math.sin(endAngle);

    const irx = innerRadiusX / outerRadiusX;
    const iry = innerRadiusY / outerRadiusY;

    const largeArc = endAngle - startAngle > Math.PI ? 1 : 0;

    paths.move(sx, sy);
    paths.arc(outerRadiusX, outerRadiusY, ex, ey, {largeArc, sweep: 1});
    paths.line(irx * ex, iry * ey);
    paths.arc(innerRadiusX, innerRadiusY, sx * irx, sy * iry, {
      largeArc,
      sweep: 0,
    });
    paths.close();

    return paths;
  }

  _checkRadius() {
    if (!isDefined(this._outerRadiusX)) {
      throw new Error('outerRadiusX must be set');
    }
  }
}

const arcFunc = () => new Arc();

export default arcFunc;
