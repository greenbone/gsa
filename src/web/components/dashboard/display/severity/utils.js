/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param {Object} params - The filter values.
 * @param {(string|number)} params.start - The start value of the filter range.
 * @param {(string|number)} params.end - The end value of the filter range.
 * @returns {FilterTerm[]} An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({start, end}) => {
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
