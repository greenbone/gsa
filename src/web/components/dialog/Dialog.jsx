/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {Modal} from '@greenbone/opensight-ui-components-mantinev7';
import {ScrollArea} from '@mantine/core';
import {isDefined, isFunction} from 'gmp/utils/identity';
import {useCallback, useEffect, useState} from 'react';
import styled from 'styled-components';
import PropTypes from 'web/utils/PropTypes';

const INITIAL_POSITION_X = 0;
const INITIAL_POSITION_Y = 70;
const MODAL_Z_INDEX = 201;
const MODAL_HEIGHT = '250px';
const MODAL_WIDTH = '40vw';

const DialogTitleButton = styled.button`
  background: none;
  border: none;
  padding: 0;
  outline: none;
  cursor: ${({$isDragging}) => ($isDragging ? 'grabbing' : 'grab')};
  display: flex;
  align-items: center;
  width: 100%;
`;

const StyledModal = styled(Modal)`
  position: relative;
  left: ${({position}) => `${position.x}px`};
  z-index: ${MODAL_Z_INDEX};

  .mantine-Modal-content {
    display: flex;
    flex-direction: column;
    position: relative;
    resize: both;
    max-height: 90vh;
    min-width: 500px;
    min-height: ${({height}) => height};
    width: ${({width}) => width};
  }

  .mantine-Modal-title {
    flex: 1;
  }
  .mantine-Modal-body {
    flex: 1;
    display: flex;
    flex-direction: column;
    overflow: hidden;
  }
  .mantine-Modal-close {
    width: 2rem;
    height: 2rem;
  }
`;

const StyledScrollArea = styled(ScrollArea)`
  flex: 1;
  overflow-y: auto;
  padding-right: 18px;
`;

const Dialog = ({
  children,
  title,
  footer,
  onClose,
  height = MODAL_HEIGHT,
  width = MODAL_WIDTH,
  testId = 'dialog-test-id',
}) => {
  const [isResizing, setIsResizing] = useState(false);

  const handleClose = useCallback(() => {
    if (isDefined(onClose)) {
      onClose();
    }
  }, [onClose]);

  const [position, setPosition] = useState({
    x: INITIAL_POSITION_X,
    y: INITIAL_POSITION_Y,
  });

  const [isDragging, setIsDragging] = useState(false);
  const [dragStart, setDragStart] = useState({x: 0, y: 0});

  useEffect(() => {
    if (isDragging) {
      document.body.style.cursor = 'grabbing';
    } else {
      document.body.style.cursor = 'default';
    }
  }, [isDragging]);

  const handleDragMouseDown = e => {
    if (
      e.target.closest('button') &&
      !e.target.closest('.dialog-title-button')
    ) {
      return;
    }
    e.preventDefault();
    setIsDragging(true);
    setDragStart({
      x: e.clientX - position.x,
      y: e.clientY - position.y,
    });
  };

  const handleMouseMove = e => {
    if (isDragging) {
      setPosition({
        x: e.clientX - dragStart.x,
        y: e.clientY - dragStart.y,
      });
    }
  };

  const handleMouseUp = () => {
    if (isResizing) {
      setIsResizing(false);
    }
    if (isDragging) {
      setIsDragging(false);
    }
  };

  return (
    <StyledModal
      centered={false}
      data-testid={testId}
      height={height}
      opened={true}
      position={position}
      size="auto"
      title={
        <DialogTitleButton
          $isDragging={isDragging}
          className="dialog-title-button"
          type="button"
          onKeyDown={e => {
            if (e.key === 'Enter' || e.key === ' ') {
              e.preventDefault();
              handleDragMouseDown(e);
            }
          }}
          onMouseDown={handleDragMouseDown}
        >
          {title}
        </DialogTitleButton>
      }
      width={width}
      yOffset={position.y}
      onClose={handleClose}
      onMouseMove={handleMouseMove}
      onMouseUp={handleMouseUp}
    >
      <StyledScrollArea>
        {isFunction(children)
          ? children({
              close: handleClose,
            })
          : children}
      </StyledScrollArea>
      {footer}
    </StyledModal>
  );
};

Dialog.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  testId: PropTypes.string,
  footer: PropTypes.node,
  title: PropTypes.string,
  width: PropTypes.string,
  onClose: PropTypes.func,
  height: PropTypes.string,
};

export default Dialog;
