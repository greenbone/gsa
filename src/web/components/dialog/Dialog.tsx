/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {useCallback, useEffect, useLayoutEffect, useState} from 'react';
import {ScrollArea} from '@mantine/core';
import {Modal} from '@greenbone/ui-lib';
import styled from 'styled-components';
import {isDefined, isFunction} from 'gmp/utils/identity';

interface DialogTitleButtonProps {
  $isDragging: boolean;
}

interface StyledModalProps {
  position: {x: number; y: number};
  height: string;
  width: string;
}

interface DialogRenderProps {
  close: () => void;
}

interface DialogProps {
  children?: React.ReactNode | ((props: DialogRenderProps) => React.ReactNode);
  title?: string;
  footer?: React.ReactNode;
  onClose?: () => void;
  height?: string;
  width?: string;
  testId?: string;
}

const INITIAL_POSITION_X = 0;
const INITIAL_POSITION_Y = 70;
const MODAL_Z_INDEX = 201;
const MODAL_HEIGHT = '250px';
const MODAL_WIDTH = '40vw';

const DialogTitleButton = styled.button<DialogTitleButtonProps>`
  background: none;
  border: none;
  padding: 0;
  outline: none;
  cursor: ${({$isDragging}) => ($isDragging ? 'grabbing' : 'grab')};
  display: flex;
  align-items: center;
  width: 100%;
`;

const StyledModal = styled(Modal)<StyledModalProps>`
  position: relative;
  left: ${({position}) => `${position.x}px`};
  z-index: ${MODAL_Z_INDEX};

  .mantine-Modal-content {
    display: flex;
    flex-direction: column;
    position: relative;
    resize: both;
    max-height: 90vh;
    min-width: 500px;
    min-height: ${({height}) => height};
    width: ${({width}) => width};
  }

  .mantine-Modal-title {
    flex: 1;
  }
  .mantine-Modal-body {
    flex: 1;
    display: flex;
    flex-direction: column;
    overflow: hidden;
  }
  .mantine-Modal-close {
    width: 2rem;
    height: 2rem;
  }
`;

const StyledScrollArea = styled(ScrollArea)`
  flex: 1;
  overflow-y: auto;
  padding-right: 18px;
`;

const Dialog = ({
  children,
  title,
  footer,
  onClose,
  height = MODAL_HEIGHT,
  width = MODAL_WIDTH,
  testId = 'dialog-test-id',
}: DialogProps) => {
  const [isResizing, setIsResizing] = useState(false);

  const handleClose = useCallback(() => {
    if (isDefined(onClose)) {
      onClose();
    }
  }, [onClose]);

  const [position, setPosition] = useState({
    x: INITIAL_POSITION_X,
    y: INITIAL_POSITION_Y,
  });

  const [isDragging, setIsDragging] = useState(false);
  const [dragStart, setDragStart] = useState({x: 0, y: 0});

  useEffect(() => {
    if (isDragging) {
      document.body.style.cursor = 'grabbing';
    } else {
      document.body.style.cursor = 'default';
    }
  }, [isDragging]);

  useLayoutEffect(() => {
    const applyStyles = () => {
      const modalRoots = document.querySelectorAll(
        '[data-portal="true"] > div > .mantine-Modal-root',
      );
      modalRoots.forEach((modalRoot, index) => {
        const overlay = modalRoot.querySelector<HTMLElement>(
          '.mantine-Modal-overlay',
        );
        if (overlay) {
          const isTopmostModal = index === modalRoots.length - 1;
          overlay.style.backgroundColor = isTopmostModal
            ? 'rgba(0, 0, 0, 0.6)'
            : 'transparent';
        }
      });
    };

    applyStyles();

    const observer = new MutationObserver(() => {
      const portals = document.querySelectorAll('[data-portal="true"]');
      if (portals.length > 0) {
        applyStyles();
      }
    });

    observer.observe(document.body, {
      childList: true,
      subtree: true,
      attributes: true,
      attributeFilter: ['class', 'style'],
    });

    return () => observer.disconnect();
  }, []);

  const handleDragMouseDown = e => {
    if (
      e.target.closest('button') &&
      !e.target.closest('.dialog-title-button')
    ) {
      return;
    }
    e.preventDefault();
    setIsDragging(true);
    setDragStart({
      x: e.clientX - position.x,
      y: e.clientY - position.y,
    });
  };

  const handleMouseMove = e => {
    if (isDragging) {
      setPosition({
        x: e.clientX - dragStart.x,
        y: e.clientY - dragStart.y,
      });
    }
  };

  const handleMouseUp = () => {
    if (isResizing) {
      setIsResizing(false);
    }
    if (isDragging) {
      setIsDragging(false);
    }
  };

  return (
    <StyledModal
      centered={false}
      data-testid={testId}
      height={height}
      opened={true}
      position={position}
      size="auto"
      title={
        <DialogTitleButton
          $isDragging={isDragging}
          className="dialog-title-button"
          type="button"
          onKeyDown={e => {
            if (e.key === 'Enter' || e.key === ' ') {
              e.preventDefault();
              handleDragMouseDown(e);
            }
          }}
          onMouseDown={handleDragMouseDown}
        >
          {title}
        </DialogTitleButton>
      }
      width={width}
      yOffset={position.y}
      onClose={handleClose}
      onMouseMove={handleMouseMove}
      onMouseUp={handleMouseUp}
    >
      <StyledScrollArea>
        {isFunction(children)
          ? children({
              close: handleClose,
            })
          : children}
      </StyledScrollArea>
      {footer}
    </StyledModal>
  );
};

export default Dialog;
