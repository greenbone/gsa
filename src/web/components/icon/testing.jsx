/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */


import {test, expect, testing} from '@gsa/testing';

import {render, fireEvent, act} from 'web/utils/testing';
import Theme from 'web/utils/theme';

import {ICON_SIZE_SMALL_PIXELS} from 'web/hooks/useIconSize';

export const testIcon = Icon => {
  test('should render with default width and height', () => {
    const {element} = render(<Icon />);

    expect(element).toHaveStyleRule('width', ICON_SIZE_SMALL_PIXELS);
    expect(element).toHaveStyleRule('height', ICON_SIZE_SMALL_PIXELS);
  });

  test('should handle click', () => {
    const handler = testing.fn();
    const {element} = render(<Icon onClick={handler} value="1" />);

    fireEvent.click(element);

    expect(handler).toHaveBeenCalledWith('1');
  });
  test('should change appearance when disabled', () => {
    const {element} = render(<Icon disabled={true} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: 'svg path',
    });
  });

  test('should change appearance when inactive', () => {
    const {element} = render(<Icon active={false} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: '& svg path',
    });
  });

  test('should not call clickhandler when disabled', () => {
    const handler = testing.fn();
    const {element} = render(<Icon disabled={true} onClick={handler} />);

    fireEvent.click(element);

    expect(handler).not.toHaveBeenCalled();
  });

  test('should render loading state', async () => {
    let resolve;
    const promise = new Promise((res, rej) => {
      resolve = res;
    });

    const handler = testing.fn().mockReturnValue(promise);

    const {element} = render(
      <Icon title="foo" loadingTitle="bar" onClick={handler} />,
    );

    const svgElement = element.querySelector('svg');

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');

    fireEvent.click(element);

    expect(element).toHaveStyleRule('display', 'none', {
      modifier: '& svg path',
    });
    expect(element).toHaveStyleRule(
      'background',
      'url(/img/loading.gif) center center no-repeat',
      {
        modifier: '& svg',
      },
    );

    expect(element).toHaveAttribute('title', 'bar');
    expect(svgElement).toHaveAttribute('title', 'bar');

    await act(async () => {
      resolve();
    });

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');
  });
};

// vim: set ts=2 sw=2 tw=80:
