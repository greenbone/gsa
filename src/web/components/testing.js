/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {isDefined} from 'gmp/utils/identity';

import {
  userEvent,
  act,
  fireEvent,
  queryAllByRole,
  queryByRole,
  getByRole,
  getAllByTestId,
  getByTestId,
} from 'web/utils/testing';

export const getElementOrDocument = element =>
  isDefined(element) ? element : document.body;

/**
 * Get all the radio inputs
 */
export const getRadioInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Radio-radio');
};

/**
 * Get all the items of a Select component
 */
export const getSelectItemElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll("[role='option']");
};

/**
 * Get the items that are selectable of a Select component
 */
export const getSelectItemElementsForSelect = async element => {
  element = isDefined(element) ? element : getSelectElement();
  await openSelectElement(element);
  const selectItemsId = element.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector('#' + selectItemsId);
  return getSelectItemElements(itemsContainer);
};

/**
 * Get the input box of a Select component
 */
export const getSelectElement = element => {
  element = getElementOrDocument(element);
  const select = queryByRole(element, 'searchbox');
  if (select) {
    return select;
  }
  return getByRole(element, 'textbox');
};

/**
 * Get all select components
 */
export const getSelectElements = element => {
  element = getElementOrDocument(element);
  return queryAllByRole(element, 'searchbox');
};

/**
 * Open a select element (MultiSelect, Select, etc.)
 */
export const openSelectElement = async select => {
  await act(async () => {
    select = select || getSelectElement();
    await userEvent.click(select);
  });
};

/**
 * Click on an element/item/node
 */
export const clickElement = async element => {
  await act(async () => {
    await userEvent.click(element);
  });
};

/**
 * Get all multi select (root) elements
 */
export const getMultiSelectElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-root');
};

/**
 * Get all multi select (root) element
 */
export const getMultiSelectElement = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.mantine-MultiSelect-root');
};

/**
 * Open a MultiSelect component for displaying item selection
 */
export const openMultiSelectElement = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  const input = getSelectElement(element);
  return await clickElement(input);
};

/**
 * Get current selected items of a MultiSelect component
 */
export const getSelectedItems = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-value');
};

/**
 * Get all the selectable items of a MultiSelect component
 */
export const getSelectItemElementsForMultiSelect = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  await openMultiSelectElement(element);
  const controlElement = element.querySelector('.mantine-MultiSelect-wrapper');
  const selectItemsId = controlElement.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector(
    '#' + selectItemsId + '-items',
  );
  return getSelectItemElements(itemsContainer);
};

/**
 * Change the value of an input element like Select or TestField component
 */
export const changeInputValue = (element, value) => {
  fireEvent.change(element, {target: {value}});
};

/**
 * Change the input of a Select component
 */
export const changeSelectInput = (value, input) => {
  if (!isDefined(input)) {
    input = getSelectElement();
  }
  changeInputValue(input, value);
};

/**
 * Query if a dialog is present
 */
export const queryDialog = element => {
  element = getElementOrDocument(element);
  return queryByRole(element, 'dialog');
};

/**
 * Get a dialog
 */
export const getDialog = element => {
  element = getElementOrDocument(element);
  return getByRole(element, 'dialog');
};

/**
 * Get the dialog content
 */
export const getDialogContent = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-body');
};

/**
 * Get the dialog title
 */
export const getDialogTitle = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-title');
};

/**
 * Get the save button of a dialog
 */
export const getDialogSaveButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-save-button');
};

/**
 * Get the close button of a dialog (in the footer)
 */
export const getDialogCloseButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-close-button');
};

/**
 * Close a dialog
 */
export const closeDialog = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  const closeButton = dialog.querySelector('.mantine-CloseButton-root');
  fireEvent.click(closeButton);
};

/**
 * Get the element containing the powerfilter
 */
export const getPowerFilter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.powerfilter');
};

/**
 * Get text inputs
 */
export const getTextInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-TextInput-input');
};

/**
 * Get the table element
 */
export const getTable = element => {
  element = getElementOrDocument(element);
  return element.querySelector('table');
};

/**
 * Get the table body element
 */
export const getTableBody = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tbody');
};

/**
 * Get the table footer element
 */
export const getTableFooter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tfoot');
};

/**
 * Get the table header element
 */
export const getTableHeader = element => {
  element = getElementOrDocument(element);
  return element.querySelector('thead');
};

/**
 * Get action items
 */
export const getActionItems = element => {
  element = getElementOrDocument(element);
  return getAllByTestId(element, 'svg-icon');
};

/**
 * Get the bulk action items of a page
 */
export const getBulkActionItems = element => {
  const tableFooter = getTableFooter(element);
  return getActionItems(tableFooter);
};

/**
 * Get the check boxes
 */
export const getCheckBoxes = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Checkbox-input');
};
