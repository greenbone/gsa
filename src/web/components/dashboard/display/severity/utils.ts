/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

interface FilterValueToFilterTermsParams {
  start?: number;
  end?: number;
}

type FilterValueToFilterTermsReturnType = [FilterTerm?, FilterTerm?];

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param params - The filter values.
 * @param params.start - The start value of the filter range.
 * @param params.end - The end value of the filter range.
 * @returns An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({
  start,
  end,
}: FilterValueToFilterTermsParams): FilterValueToFilterTermsReturnType => {
  if (!isDefined(start)) {
    return [];
  }
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
