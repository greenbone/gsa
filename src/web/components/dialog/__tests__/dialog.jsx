/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {describe, test, expect, testing} from '@gsa/testing';

import {isFunction} from 'gmp/utils/identity';
import {KeyCode} from 'gmp/utils/event';

import {render, fireEvent} from 'web/utils/testing';

import Dialog from '../dialog';

describe('Dialog component tests', () => {
  test('should render a Dialog', () => {
    const handleClose = testing.fn();
    const renderFunc = testing.fn().mockReturnValue(<div />);

    const {baseElement} = render(
      <Dialog onClose={handleClose}>{renderFunc}</Dialog>,
    );

    expect(baseElement).toMatchSnapshot();

    expect(renderFunc).toHaveBeenCalled();

    // eslint-disable-next-line prefer-destructuring
    const renderProps = renderFunc.mock.calls[0][0];
    expect(isFunction(renderProps.close)).toEqual(true);
    expect(renderProps.moveProps).toBeDefined();
    expect(isFunction(renderProps.moveProps.onMouseDown)).toEqual(true);
    expect(renderProps.heightProps).toBeDefined();
    expect(renderProps.heightProps.maxHeight).toBeDefined();
  });

  test('should close Dialog', () => {
    const handleClose = testing.fn();
    const renderFunc = testing.fn().mockReturnValue(<div />);

    render(<Dialog onClose={handleClose}>{renderFunc}</Dialog>);

    expect(renderFunc).toHaveBeenCalled();

    // eslint-disable-next-line prefer-destructuring
    const renderProps = renderFunc.mock.calls[0][0];
    expect(isFunction(renderProps.close)).toEqual(true);
    renderProps.close();

    expect(handleClose).toHaveBeenCalled();
  });

  test('should close Dialog on escape key', () => {
    const handleClose = testing.fn();
    const renderFunc = testing.fn().mockReturnValue(<div />);

    const {getByRole} = render(
      <Dialog onClose={handleClose}>{renderFunc}</Dialog>,
    );

    expect(renderFunc).toHaveBeenCalled();

    fireEvent.keyDown(getByRole('dialog'), {
      key: 'Escape',
      keyCode: KeyCode.ESC,
    });

    expect(handleClose).toHaveBeenCalled();
  });
});

// vim: set ts=2 sw=2 tw=80:
