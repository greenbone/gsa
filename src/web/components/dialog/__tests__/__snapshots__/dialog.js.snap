// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component tests should render a Dialog 1`] = `
.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10% auto;
  border: 0;
  outline: 0;
  width: 800px;
  height: px;
}

.c0 {
  position: fixed;
  font-family: Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;
  font-size: 1.1em;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 0;
  background: rgba(102, 102, 102, 0.5);
  z-index: 600;
  -webkit-transition: opacity 1s ease-in;
  transition: opacity 1s ease-in;
  width: 100%;
  height: 100%;
}

.c3 {
  width: 0;
  height: 0;
  cursor: nwse-resize;
  border-right: 20px solid transparent;
  border-bottom: 20px solid transparent;
  border-top: 20px solid #fff;
}

.c2 {
  position: absolute;
  bottom: 3px;
  right: 3px;
  width: 20px;
  height: 20px;
  background: repeating-linear-gradient(     -45deg,     transparent,     transparent 2px,     #000 2px,     #000 3px   );
}

<body>
  <div
    id="portals"
  >
    <div>
      <div
        class="c0"
      >
        <div
          class="c1"
          height="px"
          role="dialog"
          tabindex="1"
          width="800px"
        >
          <div />
          <div
            class="c2"
          >
            <div
              class="c3"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
